libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,892 @@
+    }
+    return result;
+}
+
+// run the 2-key OSD simulation process, this involves using the power and mode (wifi) buttons
+// to cycle through options. unfortunately these are one-way requests so we need to use delays
+// to make sure that the camera obeys
+void AP_RunCam::handle_2_key_simulation_process(Event ev)
+{
+#if RUNCAM_DEBUG
+    if (_in_menu > 0 && ev != Event::NONE) {
+        debug("E:%d,M:%d,V:%d\n", int(ev), _in_menu, _video_recording);
+    }
+#endif
+    switch (ev) {
+    case Event::ENTER_MENU:
+        enter_2_key_osd_menu();
+        break;
+
+    case Event::EXIT_MENU:
+        // keep changing mode until we are fully out of the menu
+        if (_in_menu-- > 0) {
+            simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_MODE);
+            set_mode_change_timeout();
+            _state = State::EXITING_MENU;
+        }
+        else {
+            exit_2_key_osd_menu();
+        }
+        break;
+
+    case Event::IN_MENU_ENTER:
+        simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_SIMULATE_WIFI_BTN);  // change setting
+        set_button_press_timeout();
+        // in a sub-menu and save-and-exit was selected
+        if (_in_menu > 1 && _sub_menu_pos == (_sub_menu_lengths[_top_menu_pos] - 1)) {
+            _in_menu--;
+        // in the top-menu and save-and-exit was selected
+        } else if (_in_menu == 1 && _top_menu_pos == (RUNCAM_TOP_MENU_LENGTH - 1)) {
+            _in_menu--;
+            _state = State::EXITING_MENU;
+        } else {
+            _in_menu = MIN(_in_menu + 1, RUNCAM_OSD_MENU_DEPTH);
+        }
+        break;
+
+    case Event::IN_MENU_UP:
+    case Event::IN_MENU_DOWN:
+        simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_SIMULATE_POWER_BTN);    // move to setting
+        set_button_press_timeout();
+        if (_in_menu > 1) {
+            // in a sub-menu, keep track of the selected position
+            _sub_menu_pos = (_sub_menu_pos + 1) % _sub_menu_lengths[_top_menu_pos];
+        } else {
+            // in the top-menu, keep track of the selected position
+            _top_menu_pos = (_top_menu_pos + 1) % RUNCAM_TOP_MENU_LENGTH;
+        }
+        break;
+
+    case Event::IN_MENU_EXIT:
+        // if we are in a sub-menu this will move us out, if we are in the root menu this will
+        // exit causing the state machine to get out of sync. the OSD menu hierachy is consistently
+        // 2 deep so we can count and be reasonably confident of where we are.
+        // the only exception is if someone hits save and exit on the root menu - then we are lost.
+        if (_in_menu-- > 0) {
+            simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_MODE);     // move up/out a menu
+            set_mode_change_timeout();
+        }
+        // no longer in the menu so trigger the OSD re-enablement
+        if (_in_menu <= 0) {
+            _state = State::EXITING_MENU;
+        }
+        break;
+
+    default:
```

Comment:
Are not all events handled here?

Adding `default` means that if a new state is added it may get missed in handling somewhere.  If some state has no handling required, better to add an empty case for it.

This comment also applies to other `switch`es

Reply:
Fixed